File: BotsSDK/korebot/src/main/java/kore/botssdk/audiocodes/webrtcclient/General/ImageUtils.java
package kore.botssdk.audiocodes.webrtcclient.General;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.Rect;
import android.net.Uri;

import java.io.FileNotFoundException;
import java.io.InputStream;

public class ImageUtils {
    private static String TAG="ImageUtils";
    public static Bitmap getContactBitmapFromURI(Context context, Uri uri) {
        try {
            InputStream input = context.getContentResolver().openInputStream(uri);
            if (input == null) {
                return null;
            }
            return BitmapFactory.decodeStream(input);
        } catch (FileNotFoundException e) {
            Log.e(TAG,"error: "+e);

        }
        return null;
    }

    public static Bitmap getCroppedRoundBitmap(Bitmap bmp, int radius) {
        Bitmap resizedBitmap = Bitmap.createScaledBitmap(bmp, radius, radius, false);
        return getCroppedRoundBitmap(resizedBitmap);
    }

    public static Bitmap getCroppedRoundBitmap(Bitmap bmp) {
        Bitmap bitmap = bmp.copy(Bitmap.Config.ARGB_8888, true);
        int w = bmp.getWidth();
        int h = bmp.getHeight();
        int radius = h < w?h:w;
        Bitmap roundBitmap = getCroppedBitmap(bitmap, radius);
        return roundBitmap;
    }

    public static Bitmap getCroppedBitmap(Bitmap bmp, int radius) {
        bmp = cropToSquare(bmp);
        Bitmap sbmp;
        if(bmp.getWidth() == radius && bmp.getHeight() == radius) {
            sbmp = bmp;
        } else {
            sbmp = Bitmap.createScaledBitmap(bmp, radius, radius, false);
        }

        Bitmap output = Bitmap.createBitmap(sbmp.getWidth(), sbmp.getHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(output);
        int color = -6187148;
        Paint paint = new Paint();
        Rect rect = new Rect(0, 0, sbmp.getWidth(), sbmp.getHeight());
        paint.setAntiAlias(true);
        paint.setFilterBitmap(true);
        paint.setDither(true);
        canvas.drawARGB(0, 0, 0, 0);
        paint.setColor(Color.parseColor("#ffffff"));
        canvas.drawCircle((float)(sbmp.getWidth() / 2), (float)(sbmp.getHeight() / 2), (float)(sbmp.getWidth() / 2), paint);
        paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC_IN));
        canvas.drawBitmap(sbmp, rect, rect, paint);
        return output;
    }

    public static Bitmap cropToSquare(Bitmap bitmap) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();
        int newWidth = height > width?width:height;
        int newHeight = height > width?height - (height - width):height;
        int cropW = (width - height) / 2;
        cropW = cropW < 0?0:cropW;
        int cropH = (height - width) / 2;
        cropH = cropH < 0?0:cropH;
        Bitmap cropImg = Bitmap.createBitmap(bitmap, cropW, cropH, newWidth, newHeight);
        return cropImg;
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/audiocodes/webrtcclient/general/ImageUtils.kt
package com.kore.ui.audiocodes.webrtcclient.general

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.PorterDuff
import android.graphics.PorterDuffXfermode
import android.graphics.Rect
import android.net.Uri
import androidx.core.graphics.toColorInt
import com.kore.common.utils.LogUtils
import java.io.FileNotFoundException

object ImageUtils {
    private const val TAG = "ImageUtils"
    fun getContactBitmapFromURI(context: Context, uri: Uri?): Bitmap? {
        try {
            val input = context.contentResolver.openInputStream(uri!!) ?: return null
            return BitmapFactory.decodeStream(input)
        } catch (e: FileNotFoundException) {
            LogUtils.e(TAG, "error: $e")
        }
        return null
    }

    @JvmStatic
    fun getCroppedRoundBitmap(bmp: Bitmap?, radius: Int): Bitmap {
        val resizedBitmap = Bitmap.createScaledBitmap(bmp!!, radius, radius, false)
        return getCroppedRoundBitmap(resizedBitmap)
    }

    private fun getCroppedRoundBitmap(bmp: Bitmap): Bitmap {
        val bitmap = bmp.copy(Bitmap.Config.ARGB_8888, true)
        val w = bmp.width
        val h = bmp.height
        val radius = if (h < w) h else w
        return getCroppedBitmap(bitmap, radius)
    }

    private fun getCroppedBitmap(bmp1: Bitmap, radius: Int): Bitmap {
        var bmp = bmp1
        bmp = cropToSquare(bmp)
        val sbmp: Bitmap = if (bmp.width == radius && bmp.height == radius) {
            bmp
        } else {
            Bitmap.createScaledBitmap(bmp, radius, radius, false)
        }
        val output = Bitmap.createBitmap(sbmp.width, sbmp.height, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(output)
        val paint = Paint()
        val rect = Rect(0, 0, sbmp.width, sbmp.height)
        paint.isAntiAlias = true
        paint.isFilterBitmap = true
        paint.isDither = true
        canvas.drawARGB(0, 0, 0, 0)
        paint.color = "#ffffff".toColorInt()
        canvas.drawCircle((sbmp.width / 2).toFloat(), (sbmp.height / 2).toFloat(), (sbmp.width / 2).toFloat(), paint)
        paint.setXfermode(PorterDuffXfermode(PorterDuff.Mode.SRC_IN))
        canvas.drawBitmap(sbmp, rect, rect, paint)
        return output
    }

    private fun cropToSquare(bitmap: Bitmap): Bitmap {
        val width = bitmap.width
        val height = bitmap.height
        val newWidth = if (height > width) width else height
        val newHeight = if (height > width) height - (height - width) else height
        var cropW = (width - height) / 2
        cropW = if (cropW < 0) 0 else cropW
        var cropH = (height - width) / 2
        cropH = if (cropH < 0) 0 else cropH
        return Bitmap.createBitmap(bitmap, cropW, cropH, newWidth, newHeight)
    }
}
